### PR TITLE
feat(conformance): add T005-T015 test-quality rules for coverage meaningfulness

### PR DESCRIPTION
**Pull request by cmgrote**
## Summary

Extends the T-series (test-quality) conformance rules to check that connector-app test coverage is *meaningful*, not just present — closing the gap where a coverage percentage can be reached without tests ever asserting an outcome. Encodes the agreed testing-tier architecture (unit + integration required, e2e recommended, UI optional) as deterministic, WARN-tier checks.

- **T005–T007** (assertion meaningfulness): `AssertionFreeTest`, `EmptyTestBody`, `VacuousAssertion` — catches tests that run without ever verifying an outcome, placeholder stubs, and `assert True`-style constant assertions that can never fail.
- **T008–T009** (silent non-execution): `UncollectableTestFile` (filename doesn't match pytest's collection glob — never runs), `UnconditionalModuleSkip` (unguarded `pytest.skip(allow_module_level=True)` disables an entire file).
- **T010–T013** (test-tier structure): `MissingUnitTestSuite` (mandatory, no exemption), `MissingIntegrationTestSuite`, `MissingE2ETestSuite` (both exemptable per-repo via `[tool.conformance].exempt_test_tiers` in `pyproject.toml`, since `atlan.yaml` is generated and must not be hand-edited), and `TestFileOutsideTierDir`.
- **T014–T015** (coverage-config integrity): `CoverageGateDisabled` (`fail_under` absent/0), `CoverageOmitsProductCode` (`omit`/`source` hiding real `app/` code from the coverage denominator).

All new rules ship at **WARN** tier for a soft rollout, matching the T001–T004 precedent.

## Implementation notes

- Three new check modules: `test_quality.py`, `test_structure.py`, `coverage_config.py`.
- Factored two helpers out of `integration_marking.py` into `_ast_common` for reuse: the pytest collection-mirror (`is_test_function`/`is_test_class`/`is_collectable_test_file`) and a new TOML suppression-directive parser (`parse_toml_suppressions`/`make_toml_finding`), mirroring the existing D-series pattern for `pyproject.toml`-anchored findings.
- Wired into `runner.py`; regenerated `docs/rules/tests.md` (staleness gate passes).
- Extended the `/remediate` OpenProse guidance in `programs/areas/tests.prose.md` with a fix prescription per new rule.
- Updated `docs/standards/testing.md` and `docs/agents/testing.md`.
- Found and fixed a real false-positive during verification: the canonical e2e reference pattern (a `Test*` class inheriting all its test methods from an SDK-generated base class, defining none of its own) was misdetected as "no tests" by T012. Fixed with a regression test, verified clean against `atlan-hello-world-app`, `atlan-openapi-app`, `atlan-mysql-app`, and `atlan-metabase-app`.

## Test plan

- [x] `uv run pytest tests/` in `packages/conformance` — 1735 passed
- [x] `uv run atlan-application-sdk-conformance gen-rule-docs --check` — docs up to date
- [x] Ran `--series T` against the four canonical apps (hello-world, openapi, mysql, metabase) — findings match expected tier gaps, no false positives after the fix
- [x] `uv run pre-commit run --all-files` — clean
- [x] `uv run python -m pytest tests/unit -x -q` — 5384 passed, 9 skipped

### CODE REVIEW

**Comment by cmgrote** [comment]
@sdk-review

**Comment by cmgrote** [comment]
## SDK Review (mothership): PR #2540 — feat(conformance): add T005-T015 test-quality rules for coverage meaningfulness

### Verdict: READY TO MERGE

> This is a well-scoped, additive conformance-suite change: eleven new WARN-tier T-series rules covering assertion meaningfulness, silent non-execution, test-tier structure, and coverage-config integrity, each paired with remediation guidance and positive/negative tests. The refactor extracting shared AST helpers into `_ast_common` is behavior-preserving and verified against the existing D-series/integration-marking tests. One doc nit found during review was fixed in a follow-up commit before this review posted.

---

### Findings

None outstanding — the sole finding from this review (below) was fixed before posting.

**`packages/conformance/conformance/docs/schema-contract.md`** — *fixed in 7b59433f*
- **Nit** [CATALOG] L220 (pre-fix) — §6.4 stated `exempt_test_tiers` is "T010–T012's key," but `test_structure.py` implements and documents T010 as non-exemptable via that config-based mechanism (`_REQUIRED_TIERS` only contains T011/T012; T010's check never reads `exempt_test_tiers`; the module docstring says "unit (T010) is never exemptable"; and this PR's own `docs/standards/testing.md` correctly scopes the exemption to "the integration or e2e tier" only). Note: T010 can still be suppressed per-instance via the generic inline `# conformance: ignore[T010] <reason>` directive — only the durable, repo-wide `exempt_test_tiers` config key excludes it. *Fix applied: "T010–T012's key" → "T011–T012's key".*

### Strengths
- Every new rule (T005–T015) ships at WARN tier and is dogfooded correctly — the catalog test asserts an exact rule set (not a subset), so a scope mistake would fail CI on its own.
- The claimed T012 (`MissingE2ETestSuite`) false-positive fix — an e2e subclass with no own test methods, inheriting all tests from a base class — is backed by a real regression test, plus a second test documenting the one remaining known gap.
- The `_ast_common` extraction (`is_test_function`/`is_test_class`/`is_collectable_test_file`, `parse_toml_suppressions`/`make_toml_finding`) is additive plus a behavior-preserving refactor of `integration_marking.py` and `dependency_conformance.py`, verified by the full affected test suite (213 tests) passing and `gen-rule-docs --check` confirming docs are current.
- Rule + remediation shipped together: every T005–T015 rule has a matching entry in `tests.prose.md` with explicit fix-classification guidance.
- T014/T015 coverage-config checks correctly account for CI-declared `fail_under` overriding `pyproject.toml`, with conservative, tested false-positive scoping on `omit`/`source`.
- CI workflow diff (`.github/workflows/conformance-reusable.yaml`) is a comment + matrix-glob widening only — no new injection surface, permissions/pinning untouched.

---
**CI:** all passing (sdk-review itself pending — this review)
**Models:** Claude Sonnet 5 (review agents: conformance-suite specialist + ci-config specialist). GPT-5.3-codex adversarial pass was unavailable — no proxy access in this interactive session — so findings were kept at the reviewing agent's own confidence threshold (≥0.80) per the de-bias fallback rule.
**Run:** interactive Claude Code session (not a mothership dispatch) — no GHA run URL applies.

File: packages/conformance/conformance/suite/checks/test_quality.py [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+        anchor = collected[0][0]
+        return [
+            make_finding(
+                filename=file,
+                rule_id=RULE_T008,
+                node=anchor,
+                message=_t008_message(file),
+                directives=directives,
+            )
+        ]
```
[the remark this answers is not in the document]
Comment (cmgrote) [reply packages/conformance/conformance/suite/checks/test_quality.py:392]: Fixed in 7412cbdc — the finding is now anchored via `node=tree` (the module AST), and `make_finding` falls back to `getattr(node, "lineno", 1)`; since `ast.Module` has no `lineno` attribute, this resolves to line 1. Verified: `# conformance: ignore[T008] ...` on the file's first line now suppresses the finding regardless of where the first collected test appears.

File: packages/conformance/conformance/suite/checks/coverage_config.py [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+# layouts — anything else rooted there is real product code.
+_LEGIT_OMIT_MARKERS: tuple[str, ...] = ("generated",)
+
+
+def _is_legit_app_omit(pattern: str) -> bool:
+    norm = pattern.strip().strip("\"'").replace("\\", "/")
+    if any(marker in norm for marker in _LEGIT_OMIT_MARKERS):
+        return True
+    tail = norm.rsplit("/", 1)[-1]
+    return tail.startswith(("test_", "conftest")) or tail.endswith("_test.py")
```
[the remark this answers is not in the document]
Comment (cmgrote) [reply packages/conformance/conformance/suite/checks/coverage_config.py:199]: Fixed in 7412cbdc — `_is_legit_app_omit` now splits the pattern into path segments and checks `"generated" in segments`, instead of a substring check. Verified: `app/generated_helpers/*` and `app/not_generated/*` no longer match (segments are `generated_helpers`/`not_generated`, neither equals `generated`), while `app/generated/*` still matches correctly.

File: packages/conformance/tests/test_test_structure.py [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+            "pyproject.toml": (
+                '[tool.conformance]\nexempt_test_tiers = ["integration", "e2e"]\n'
+            ),
+            "tests/unit/test_foo.py": _ONE_TEST,
+        },
+    )
+    assert _run(tmp_path) == []
+
+
+def test_partial_exemption_still_fires_for_non_exempt_tier(tmp_path: Path) -> None:
```
[the remark this answers is not in the document]
Comment (cmgrote) [reply packages/conformance/tests/test_test_structure.py:164]: Fixed in 7412cbdc — the test's fixture no longer includes `tests/integration/test_bar.py`, so the integration tier is genuinely missing, and the assertion now reads `assert _ids(_run(tmp_path)) == ["T011"]`, matching the test name ("still fires for non-exempt tier").
